core/src/main/java/tech/pegasys/eth2signer/core/signing/LoadedSigners.java

```diff
@@ -0,0 +1,111 @@
+import tech.pegasys.eth2signer.core.multikey.SignerLoader;
+import tech.pegasys.eth2signer.core.multikey.metadata.BlsArtifactSignerFactory;
+import tech.pegasys.eth2signer.core.multikey.metadata.Secp256k1ArtifactSignerFactory;
+import tech.pegasys.eth2signer.core.multikey.metadata.parser.YamlSignerParser;
+import tech.pegasys.signers.hashicorp.HashicorpConnectionFactory;
+import tech.pegasys.signers.secp256k1.azure.AzureKeyVaultSignerFactory;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import io.vertx.core.Vertx;
+import org.hyperledger.besu.plugin.services.MetricsSystem;
+
+public class LoadedSigners {
+
+  private final Map<SignerTypes, List<ArtifactSigner>> signersByType;
+
+  private LoadedSigners(final Map<SignerTypes, List<ArtifactSigner>> signersByType) {
+    this.signersByType = signersByType;
+  }
+
+  public static LoadedSigners loadFrom(
+      final Config config, final Vertx vertx, final MetricsSystem metricsSystem) {
+
+    final AzureKeyVaultSignerFactory azureFactory = new AzureKeyVaultSignerFactory();
+    final HashicorpConnectionFactory hashicorpConnectionFactory =
+        new HashicorpConnectionFactory(vertx);
+
+    final BlsArtifactSignerFactory blsArtifactSignerFactory =
+        new BlsArtifactSignerFactory(
+            config.getKeyConfigPath(), metricsSystem, hashicorpConnectionFactory);
+
+    final Secp256k1ArtifactSignerFactory ethSecpArtifactSignerFactory =
+        new Secp256k1ArtifactSignerFactory(
+            hashicorpConnectionFactory,
+            config.getKeyConfigPath(),
+            azureFactory,
+            EthSecpArtifactSigner::new);
+
+    final Secp256k1ArtifactSignerFactory fcSecpArtifactSignerFactory =
+        new Secp256k1ArtifactSignerFactory(
+            hashicorpConnectionFactory,
+            config.getKeyConfigPath(),
+            azureFactory,
+            signer -> new FcSecpArtifactSigner(signer, config.getFilecoinNetwork()));
+
+    final Collection<ArtifactSigner> signers =
+        SignerLoader.load(
+            config.getKeyConfigPath(),
+            "yaml",
+            new YamlSignerParser(
+                List.of(
+                    blsArtifactSignerFactory,
+                    ethSecpArtifactSignerFactory,
+                    fcSecpArtifactSignerFactory)));
+
+    return new LoadedSigners(
+        signers
+            .parallelStream()
+            .collect(
+                Collectors.groupingBy(
+                    i -> {
+                      if (i instanceof BlsArtifactSigner) {
+                        return SignerTypes.BLS;
+                      } else if (i instanceof EthSecpArtifactSigner) {
+                        return SignerTypes.ETH_SECP;
+                      } else if (i instanceof FcSecpArtifactSigner) {
+                        return SignerTypes.FC_SECP;
+                      } else {
+                        return SignerTypes.FC_BLS;
```

Comment:
ideally we should throw an IllegalStateException if we encountered an invalid (or potentially new signer type).

Reply:
done